### PR TITLE
ENG-90762 ENG-90941 execute-esq tool

### PR DESCRIPTION
**Pull request by d-baranovskyi-creatio**
1. Added execute-esq tool that covers all capabilities of ESQ SelectQuery
2. Added guidances `esq` and `esq-filters`
3. Added ESQ usage for validating lookups in business rules conditions and set values action
4. Update descriptions and guidances for business rules to use execute-esq for resolving lookups

### CODE REVIEW

File: clio/Command/BusinessRules/BusinessRuleLookupReferenceValidator.cs [hunk 1]
```diff
@@ -145,30 +145,29 @@ private static bool TryBuildLookupReference(
...
-		if (!ResponseContainsRecord(responseJson)) {
+		// ESQ SelectQuery (DataService) is used here instead of OData because it resolves lookup records
+		// more reliably across schemas; mirrors the static-filter LookupValueResolver.
+		object query = SelectQueryHelper.BuildSelectQuery(
+			reference.ReferenceSchemaName,
+			[new SelectQueryHelper.SelectQueryColumnDefinition("Id", "Id")],
+			[new SelectQueryHelper.SelectQueryFilterDefinition(
+				"Id",
+				reference.RecordId.ToString("D"),
+				SelectQueryHelper.TextDataValueType,
```
Comment (d-baranovskyi-creatio) [inline clio/Command/BusinessRules/BusinessRuleLookupReferenceValidator.cs]: Blocker: the Id (primary-key Guid) column is filtered with `SelectQueryHelper.TextDataValueType` (1), but it must be `GuidDataValueType` (0).

Three independent sources in this same repo say so:
- This PR's own guidance, `EsqFiltersGuidanceResource.cs`: "Guid (0): value is the GUID string. Used for raw Id comparisons and on .Id leaves of backward references."
- The sibling method this code says it "mirrors", `LookupValueResolver.TryResolveDisplayNameById` (LookupValueResolver.cs:99), filters `Id` with `GuidDataValueType` (0).
- Every other Id/UId Guid filter in the codebase uses `GuidDataValueType` (ApplicationInfoService, DataBindingDbCommand, FindEntitySchemaCommand, etc.). This call site is the only outlier.

Risk: DataService compares a Guid column against a Text parameter. The likely outcome is a server-side ESQ error (success:false) -> `SelectQueryHelper.ExecuteSelectQuery` throws -> business-rule creation fails for ANY rule referencing a lookup constant by GUID, i.e. the feature's primary happy path. Even if the server silently coerces, the contract is wrong and contradicts the shipped guidance.

The existing unit tests do not catch this: they mock the HTTP response and only assert the GUID substring is present in the body, never the `dataValueType`.

Fix: use `SelectQueryHelper.GuidDataValueType` here, and add a test asserting the serialized filter carries `"dataValueType":0` for the Id filter. Please confirm against a live environment.

File: clio/Command/BusinessRules/BusinessRuleLookupReferenceValidator.cs [hunk 2]
```diff
@@ -145,30 +145,29 @@ private static bool TryBuildLookupReference(
...
+				ComparisonType: 3)]);
+
+		LookupExistsResponseDto response = SelectQueryHelper.ExecuteSelectQuery<LookupExistsResponseDto>(
+			applicationClient, serviceUrlBuilder, query);
```
Comment (d-baranovskyi-creatio) [inline clio/Command/BusinessRules/BusinessRuleLookupReferenceValidator.cs]: Major (error-handling regression): `SelectQueryHelper.ExecuteSelectQuery` throws `InvalidOperationException` on a `success:false` server envelope (SelectQueryHelper.cs:39-43). This method only translates the empty-rows case into a guidance-bearing `ArgumentException`; it does not wrap the helper call. So any server-side ESQ failure during validation propagates as a raw `InvalidOperationException`.

This breaks the module's failure contract: validation elsewhere reports problems as `ArgumentException` with actionable, agent-facing hints, and `PageBusinessRuleValidator` only catches `ArgumentException` (so its candidate-hint enrichment is skipped for this path). The previous OData implementation parsed a value array and reported not-found cleanly rather than throwing a transport-style exception, so this is a behavioral regression.

Fix: wrap the `ExecuteSelectQuery` call and translate `InvalidOperationException` into an `ArgumentException` using the same guidance message style as the not-found branch below. (Fixing the dataValueType issue above removes the most common trigger but not the general case.)

File: clio/Command/McpServer/Tools/ToolContractGetTool.cs [hunk 3]
```diff
@@ -289,6 +289,7 @@ internal static class ToolContractCatalog {
 		new Dictionary<string, ToolContractDefinition>(StringComparer.OrdinalIgnoreCase) {
 			[ToolContractGetTool.ToolName] = BuildToolContractGet(),
 			[GuidanceGetTool.ToolName] = BuildGuidanceGet(),
+			[ExecuteEsqTool.ToolName] = BuildExecuteEsq(),
```
Comment (d-baranovskyi-creatio) [inline clio/Command/McpServer/Tools/ToolContractGetTool.cs:292]: Major (discoverability): `execute-esq` is added to the `Contracts` dictionary here, but it is NOT added to `CanonicalToolNames` (lines ~353-400). `GetContracts(null)` returns only `CanonicalToolNames`, so a caller listing the default contract set will not discover `execute-esq` — it is only reachable if the caller already knows the exact name and passes it in `tool-names`.

Every comparable read tool (`odata-read`, `page-list`, `find-entity-schema`, `get-guidance`) is in `CanonicalToolNames`. The parent story (ENG-90762) explicitly wants the agent to validate an entity/column/filter via execute-esq before saving a metric widget, so hiding it from the canonical discovery set works against the feature's purpose.

Fix: add `ExecuteEsqTool.ToolName` to `CanonicalToolNames`. If the omission is intentional, add a comment explaining why, since it diverges from every other read tool.

File: clio/Command/McpServer/Tools/ExecuteEsqTool.cs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+			// Explicit DataService failure envelope.
+			if (root.TryGetProperty("success", out JsonElement successEl)
+				&& successEl.ValueKind == JsonValueKind.False) {
+				return ExecuteEsqResponse.Failure(ExtractErrorMessage(root) ?? Truncate(json));
+			}
+
+			// Error bodies that carry no rows: DataService responseStatus / errorInfo, or an ASP.NET error body.
+			bool hasRows = root.TryGetProperty("rows", out JsonElement rowsEl);
+			if (!hasRows
```
Comment (d-baranovskyi-creatio) [inline clio/Command/McpServer/Tools/ExecuteEsqTool.cs]: Major (correctness): This no-rows error branch does not re-check `success`. A SelectQuery response with `success:true` but no `rows` array that also carries a top-level `responseStatus` / `Message` / `errorInfo` (a non-row projection, or a success envelope that still includes a `responseStatus` block) is funneled into `ExecuteEsqResponse.Failure(...)`, surfacing `ExtractErrorMessage(...) ?? Truncate(json)` as an error on a successful call.

Gate this branch on `success` not being explicitly true. Capture it once near line 105, e.g. `bool successTrue = root.TryGetProperty("success", out var s) && s.ValueKind == JsonValueKind.True;`, and only treat the error markers as a failure when `!successTrue`. No test currently covers a success body that carries these markers, so add one alongside the fix.

File: clio/Command/McpServer/Tools/ToolContractGetTool.cs [hunk 5]
```diff
@@ -351,6 +352,7 @@ internal static class ToolContractCatalog {
 
 	private static readonly string[] CanonicalToolNames = [
 		GuidanceGetTool.ToolName,
+		ExecuteEsqTool.ToolName,
```
Comment (d-baranovskyi-creatio) [inline clio/Command/McpServer/Tools/ToolContractGetTool.cs:355]: Major (test gap): Adding `execute-esq` to `CanonicalToolNames` here (and to the `Contracts` dictionary at line 292) correctly resolves the discoverability issue from the prior review — but the fix has no regression test. `GetContracts(null)` returning the canonical set is exactly what callers rely on at bootstrap, and a dropped `Contracts[name]` entry would throw `KeyNotFoundException` for every bootstrap call. There is currently no test referencing `execute-esq` in `ToolContractGetToolTests`.

Add: (1) a test asserting the canonical contract set (empty request) contains `execute-esq`; (2) a per-contract test for `BuildExecuteEsq` covering required inputs (`query`, `environment-name`), read-only intent, the `rows`/`count`/`success` output contract, and the get-guidance preferred flow.

File: clio.tests/Command/McpServer/ExecuteEsqToolTests.cs [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+using Clio.Command.McpServer.Tools;
+using Clio.Common;
+using FluentAssertions;
+using ModelContextProtocol.Server;
+using NSubstitute;
+using NUnit.Framework;
+
+namespace Clio.Tests.Command.McpServer;
+
+[TestFixture]
```
Comment (d-baranovskyi-creatio) [inline clio.tests/Command/McpServer/ExecuteEsqToolTests.cs:14]: Major (convention / test selection): This fixture is missing `[Property("Module", "McpServer")]`. Every other MCP fixture in `clio.tests/Command/McpServer/` carries it, and AGENTS.md "Smart regression testing policy" selects targeted tests via `--filter "Category=Unit&Module=McpServer"`. As written, the entire execute-esq unit suite is silently excluded from the documented module filter, so the mandatory pre-commit gate for this module skips these tests.

Add `[Property("Module", "McpServer")]` to the fixture (below `[TestFixture]`), matching the sibling fixtures.

**Review by d-baranovskyi-creatio — COMMENTED** [review]
## PR Review (re-review) — ENG-90762 / ENG-90941: execute-esq tool

**Decision: Request changes** (posted as a COMMENT event — GitHub blocks a formal review event on one's own PR; 3 inline comments were also posted).

Multi-agent review (code-quality, security, performance, testing, bug-risk) against the latest head `12ccee9b`. The three inline findings from the prior review are all **verified fixed**: the `Id` filter now uses `GuidDataValueType` (0), the server-failure envelope is wrapped as `ArgumentException` to preserve the `PageBusinessRuleValidator` contract, and `execute-esq` is now registered in both `Contracts` and `CanonicalToolNames`. Each has a regression test except discoverability (see below). Overall the design is solid: read-only `SelectQuery` through the permission-enforcing DataService, accurate `ReadOnly`/`Destructive=false` advertisement, DI-resolved clients, MCP surface (tool + contract + 2 guidance resources + unit + e2e) present per the MCP maintenance policy.

### Findings (highest severity first)

**Major — Response parser can misclassify a successful no-rows response as a failure**
`clio/Command/McpServer/Tools/ExecuteEsqTool.cs:112-118`. The no-rows error branch does not re-check `success`. A `success:true` body with no `rows` array but a top-level `responseStatus`/`Message`/`errorInfo` is reported as an error. Gate the branch on `success` not being explicitly true. (Inline comment posted.) Confidence: logic gap is certain; how often a real Creatio success body carries these markers is medium.

**Major — No regression test for `execute-esq` discoverability in the canonical contract set**
`clio/Command/McpServer/Tools/ToolContractGetTool.cs:355` / `292`. The previously-Major discoverability fix has no guard; a dropped `Contracts` entry would throw `KeyNotFoundException` at bootstrap. Add a `GetContracts(null)`-includes-`execute-esq` test plus a `BuildExecuteEsq` per-contract test. (Inline comment posted.)

**Major — New test fixture excluded from smart test selection**
`clio.tests/Command/McpServer/ExecuteEsqToolTests.cs:13` is missing `[Property("Module", "McpServer")]`, which every sibling MCP fixture carries. Per AGENTS.md the module filter `Category=Unit&Module=McpServer` would silently skip the entire execute-esq suite in the pre-commit gate. (Inline comment posted.)

### Minor

- **Timeout mislabeled as a query-format error** — `ExecuteEsqTool.cs:61-63`. A timeout/`OperationCanceledException` is caught by the broad `catch` and returned with the "ESQ format is easy to get wrong" guidance hint, which is actively misleading for a timeout. Add a dedicated catch returning a timeout-specific message without the format hint.
- **`because` missing on several assertions** — `ExecuteEsqToolTests.cs` (e.g. the `Should().BeFalse()`/`BeTrue()` calls in `...Attach_Guidance_Hint_On_Failure`, `...Not_Attach_Hint_On_Success`, `...Fail_When_Environment_Missing`, `...Query_Is_Not_Object`, `...Empty_Response`, `...Return_Whole_Body_When_No_Rows_Array`, `...Client_Throws`). AGENTS.md makes `because` a hard rule; the file is inconsistent with its own better-written tests.
- **Uncovered parse branches** — non-object query string ("query string must contain a JSON SelectQuery object"), `responseStatus` without `ErrorCode`, `rows` present but non-array/null, and the `Failed to parse SelectQuery response` catch. Add targeted unit tests.
- **`execute-esq` has no row cap and honors `rowCount:-1`** — `ExecuteEsqTool.cs`. An AI caller that omits `rowCount` triggers an unbounded `SelectQuery`, and the full `rows` array is cloned into the response. Consider injecting a sane default cap when none is supplied.
- **Existence probe over-fetches** — `SelectQueryHelper.BuildSelectQuery` hardcodes `rowCount=10000`; the validator only needs to know if >=1 row exists. Bounded to <=1 row by the PK-equality filter so impact is minimal here, but a latent footgun for other callers; consider an optional `rowCount` parameter (validator: 1, `LookupValueResolver` ambiguity check: 2).
- **Brittle JSON-substring assertions** — the load-bearing `dataValueType:0` regression assertion in `BusinessRuleLookupReferenceValidatorTests` depends on compact `System.Text.Json` output; prefer parsing the captured body and asserting on the `JsonElement` value.
- **Duplicated error-extraction logic** — `ExecuteEsqTool.ExtractErrorMessage` vs `SelectQueryHelper.ExecuteSelectQuery` both encode "what a DataService failure body looks like." Partly intentional (different consumers), but consider a shared `JsonElement`-based error reader.
- **`get-guidance` contract "known values" list omits `esq`/`esq-filters`** — `ToolContractGetTool.cs:543`. Pre-existing partial list, but the PR's value depends on callers discovering these guides; append them or reword as illustrative.
- **e2e covers only the unhappy path** — `ExecuteEsqToolE2ETests.cs` validates advertisement and arg-binding-with-invalid-env, but no successful `SelectQuery` returning rows. If a live env is available, add a minimal success-path test; otherwise note the gap explicitly.
- **Security (low, local threat model)** — raw `ExceptionMessage`/server error bodies are echoed to the caller; acceptable for a local CLI operator, but since the consumer may be an LLM context, consider returning a generic message for ASP.NET error bodies while keeping the clean DataService validation messages.

### Requirement coverage
- Added `execute-esq` covering ESQ SelectQuery capabilities — **covered** (tool, contract, unit + e2e present; success-path e2e is the gap).
- Added `esq` and `esq-filters` guidances — **covered** (resources + catalog registration + e2e retrieval tests).
- ESQ usage for validating lookups in business-rule conditions and set-values — **covered** (validator migrated to SelectQuery with Guid filter; regression tests present).
- Updated business-rule descriptions/guidances to use `execute-esq` for resolving lookups — **covered** (ToolContractGetTool, BusinessRulesGuidanceResource, LookupValueResolver text updated).

### Confidence limits
Code-only review against the diff and surrounding source (no live Creatio environment exercised). The parse-misclassification finding's real-world trigger frequency depends on Creatio DataService success-response shapes, which were not verified against a live server. The diff is extra-large (~1.4k lines), but the bulk is two static guidance text resources that were assessed for the established pattern rather than line-by-line prose.

No approval is issued (decision is request changes, and self-approval is blocked for the PR author).

**Review by d-baranovskyi-creatio — COMMENTED** [review]
## PR Re-review (round 2) — ENG-90762 / ENG-90941: execute-esq tool

**Decision: approve (minor issues)** — posted as a COMMENT event (the PR author cannot self-approve via a formal review event). No Blocker or Major findings remain.

Re-reviewed the fix commits (`12ccee9b..22534cac`) with four parallel lenses (bug-risk, testing, code-quality, performance). Security was not re-run: round-1 cleared the tool, and this delta (parse restructure, timeout catch, row-count cap, doc text, tests) adds no new security surface.

### Round-1 findings — resolution
- **[Major] Parse misclassification** (`ExecuteEsqTool.cs`) — FIXED. `ParseSelectQueryResponse` now gates the no-rows error branch on `!successIsTrue`; a `success:true` body with a `responseStatus`/`Message` and no rows array is returned as success. A non-true `success` value is treated as failure. Verified across all envelope shapes; regression test added.
- **[Major] No discoverability regression test** (`ToolContractGetTool`) — FIXED. `execute-esq` added to the canonical-set assertion plus a new `ToolContractGet_Should_Return_ExecuteEsq_Contract` per-contract test (required inputs, rows/count/success output, get-guidance flow).
- **[Major] Missing `Module` trait** (`ExecuteEsqToolTests`) — FIXED. `[Property("Module", "McpServer")]` added; the suite now runs under the documented smart-test filter.
- **[Minor] Timeout mislabeled** — FIXED. Dedicated `catch when (ex is OperationCanceledException or TimeoutException)` returns a timeout-specific message via a new `FailureWithoutGuidance` factory (no misleading ESQ-format hint).
- **[Minor] Missing `because`** — FIXED. All assertions in `ExecuteEsqToolTests` now carry `because` (verified: 0 bare `Should()` remain).
- **[Minor] Uncovered parse branches** — FIXED. Added tests for non-object query string, `responseStatus` without `ErrorCode`, non-array `rows`, malformed response, and `success:true`+`responseStatus`+no-rows; plus (round 2) the `ExtractErrorMessage ?? Truncate(json)` fallback when a `success:false` body has no extractable error field.
- **[Minor] Existence probe over-fetch** — FIXED. `SelectQueryHelper.BuildSelectQuery` gained an optional `rowCount` (default 10000, preserving `LookupValueResolver` behavior); the business-rule existence probe now caps at `rowCount: 1`.
- **[Minor] Brittle JSON-substring assertion** — FIXED. The Guid `dataValueType` assertion now parses the captured request body and navigates to `rightExpression.parameter.dataValueType`.
- **[Minor] get-guidance known-values list** — FIXED. `esq`, `esq-filters`, `business-rules` added, with an "illustrative, not exhaustive" note.

### Round-2 findings
- **[Minor] Untested truncated-body failure branch** — FIXED in `22534cac` (test `Execute_Should_Surface_Truncated_Body_When_No_Error_Field`).

No new Blocker/Major issues were introduced by the fixes. Bug-risk, code-quality, and performance lenses confirmed the parse restructure is correct for every envelope shape, the timeout catch ordering is sound, and the `rowCount` change is a strict improvement that leaves `LookupValueResolver`'s ambiguity detection (default 10000) intact.

### Deliberately not changed (not worth doing)
- `execute-esq` honors caller `rowCount:-1` with no silent cap — intentional raw-passthrough design; the 120s timeout clamp already bounds runaway queries.
- Shared error-extraction between `ExecuteEsqTool` and `SelectQueryHelper` — the divergence is intentional (throw-based DTO vs structured JsonElement envelope); a shared helper would add coupling for little gain.
- ASP.NET `ExceptionMessage` echoed to the caller — acceptable for a local-CLI threat model where the operator owns the environment.

### Remaining gap (acknowledged, not blocking)
- **e2e success path** — `ExecuteEsqToolE2ETests` covers advertisement and arg-binding/invalid-environment, but a real SelectQuery returning rows needs a live registered Creatio environment the e2e harness lacks. Tracked as a known limitation, not a deliverable gap.

### Requirement coverage
All four PR requirements remain **covered** (execute-esq tool + contract + unit/e2e; esq/esq-filters guidances; ESQ-based lookup validation in business rules; updated business-rule descriptions/guidances).

### Validation
Targeted unit suite green: `Category=Unit&(Module=McpServer|Module=Command)` → 2303 passed / 0 failed / 11 skipped; affected fixtures (incl. the round-2 test) → all green.

The review loop has converged: all actionable findings are resolved. No approval is issued (self-approval is blocked for the PR author).

File: clio/Common/ConsoleLogger.cs [hunk 7]
```diff
@@ -83,7 +83,10 @@ private void FlushQueueCore() {
 				DebugMessage debugMessage => () => PrintDebugInternal(debugMessage.Value.ToString()),
 				var _ => throw new ArgumentOutOfRangeException()
 			};
-			action.Invoke();
+			try {
+				action.Invoke();
+			} catch (ObjectDisposedException) {
```
Comment (d-baranovskyi-creatio) [inline clio/Common/ConsoleLogger.cs:88]: It fixes unit tests.
Swallowing the error is OK, because:
- It's narrowly typed to **`ObjectDisposedException`**, not a blanket `catch (Exception)`. It won't hide formatting bugs, null refs, I/O failures, etc.
- In **production, `Console.Out` is the real console and is never disposed**, so the catch literally never triggers there — no real error is being hidden.
- The only thing that disposes `Console.Out` is the **test harness** (capturing output). Swallowing in that case is exactly right — it's a test artifact, not a failure.
- The message is recorded in `LogMessages` **before** the write attempt, so swallowing doesn't lose the logical message; it only skips writing to a dead writer.

**Review by oyermakovcreatio — APPROVED** [review]
